### PR TITLE
Less

### PR DESCRIPTION
**Pull request by rjoleary**
From Michael Pratt

### CODE REVIEW

**Comment by rjoleary** [comment]
Fixes the less part of #940

File: cmds/less/less.go [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"regexp"
+	"sync"
+
+	"github.com/nsf/termbox-go" // TODO: vendor
```
Comment (prattmic) [inline cmds/less/less.go]: TODONE?

File: cmds/less/lineio/lineio.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
+// Copyright 2018 the u-root Authors. All rights reserved
```
Comment (prattmic) [inline cmds/less/lineio/lineio.go]: I'm not familiar with u-root typical package structure, but IMO lineio and sortedmap belong in pkg/... not cmd/less/... They aren't less-specific.

File: cmds/less/main.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+//     * ^U: Scroll up one half screen full
+//
+//     Searching:
+//
+//     * /: Enter search regex (re2 syntax). Press enter to search.
+//     * n: Jump down to next search result
+//     * N: Jump up to previous search result
+//
+// Author:
+//     Michael Pratt (github.com/prattmic)
```
Comment (prattmic) [inline cmds/less/main.go]: Perhaps note this is originally based on https://github.com/prattmic/lesser? Though my repo is abandoned, so this will effectively become the canonical version,

File: cmds/less/less.go [hunk 4]
```diff
@@ -72,7 +72,7 @@ type Lesser struct {
 
 	// searchResults are the results for the current search.
 	// They should be highlighted.
-	searchResults searchResults
+	searchResults *searchResults
```
Comment (prattmic) [inline cmds/less/less.go]: Why this change?

Comment (rjoleary) [reply cmds/less/less.go]: `go vet` complains when a mutex is copied by value. `searchResults` contains a mutex in there somewhere which `go vet` is smart enough to complain about as well. The fix is to copy the pointer rather than the value, but `go vet` still seems to be complaining. I'm still looking into this...

Comment (prattmic) [reply cmds/less/less.go]: NewSearchResults also returns a copy. It should probably return a pointer anyways given the name,

Comment (rjoleary) [reply cmds/less/less.go]: Thank you! I changed a bunch of values to pointers all over the code, and it seems to work. Do you know why it was an issue to copy a mutex by value?

File: cmds/less/less.go [hunk 5]
```diff
@@ -337,12 +337,13 @@ func (s *searchResults) Prev(line int64) (searchResult, bool) {
 	return s.results[i], true
 }
 
-func (l *Lesser) search(s string) searchResults {
+func (l *Lesser) search(s string) *searchResults {
 	reg, err := regexp.Compile(s)
 	if err != nil {
 		// TODO(prattmic): display a better error
 		log.Printf("regexp failed to compile: %v", err)
-		return NewSearchResults()
```
Comment (prattmic) [inline cmds/less/less.go]: Don't change this.

File: cmds/less/less.go (hunk 4 above)
Comment (prattmic) [reply cmds/less/less.go]: It is not safe to copy a sync.Mutex once it is in use. It contains internal fields about waiters that lose meaning once copied (since the waiters are waiting on the old value). Plus it just doesn't make much sense. If you are copying a Mutex, how do other users know about the new location.

In this case, the copies were all before the first use, but the linter doesn't know that and just warns regardless.

https://godoc.org/sync#Mutex

File: cmds/less/less.go (hunk 5 above)
Comment (rjoleary) [reply cmds/less/less.go]: Done! I also changed flag to pflag because that is what u-root commands are migrating too.

**Comment by rjoleary** [comment]
I'll also get Gan to help review this for good measure. We don't normally import commands point blank like this.

File: vendor/github.com/nsf/termbox-go/syscalls.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
+// +build ignore
+
+package termbox
+
+/*
+#include <termios.h>
+#include <sys/ioctl.h>
+*/
+import "C"
```
Comment (hugelgupf) [inline vendor/github.com/nsf/termbox-go/syscalls.go]: um, cgo?

File: vendor/github.com/nsf/termbox-go/syscalls.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// +build ignore
```
Comment (hugelgupf) [inline vendor/github.com/nsf/termbox-go/syscalls.go]: wat?

File: vendor/github.com/nsf/termbox-go/syscalls_darwin.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
+// Created by cgo -godefs - DO NOT EDIT
+// cgo -godefs syscalls.go
```
Comment (hugelgupf) [inline vendor/github.com/nsf/termbox-go/syscalls_darwin.go]: oh is this generated from the cgo ignored file? weird

File: vendor/github.com/nsf/termbox-go/syscalls.go (hunk 6 above)
Comment (rjoleary) [reply vendor/github.com/nsf/termbox-go/syscalls.go]: I assure you that when you run `CGO_ENABLED=0 go build`, it generates a statically linked executable. I think this file is ignored, but dep still imports it.

File: vendor/github.com/nsf/termbox-go/syscalls.go (hunk 7 above)
Comment (rjoleary) [reply vendor/github.com/nsf/termbox-go/syscalls.go]: I don't think this is too important. If you `git grep \"C\", you'll see many other vendored packaged do this as well. There might be some way of telling dep to prune cgo source.

**Comment by rminnich** [comment]
so are we good to go with this? would be nice to have less.

**Comment by rjoleary** [comment]
I did a bit of cleanup and rebased. Feel free to review again.
